### PR TITLE
Implement testing for Multiclass_longdesc and Multiclass_shortdesc

### LINKED ISSUE #1402: Create testing for multiclass module functions multiclass_shortdesc() and multiclass_longdesc to test the new implementation
We will move previous bug fixes to updated version of the multiclass module, from class.h and class.c to multiclass.h and multiclass.c. Also, the implementations of multiclass_shortdesc() and multiclass_longdesc() have been updated to fix some bugs in issues #1142 and #1143 so tests for these functions must be written.

### PR DESCRIPTION
**Pull request by wellingtj**
The new multiclass has several functions, and they all need to be tested. This pull request implements multiclass_longdesc, multiclass_shortdesc and their tests. Currently, the tests account for grammar, string overflow, and overall functionality within the two functions. Additionally, it debugs the multiclass function and a few of the other functions in the multiclass module. This pull request is the first step in finishing issue #1608.

### CODE REVIEW

File: src/playerclass/src/multiclass.c [hunk 1]
```diff
@@ -56,11 +112,21 @@ char* multiclass_shortdesc(class_t* base_class, class_t* second_class) {
  * Returns:
```
Comment (vanolson) [inline src/playerclass/src/multiclass.c:118]: Okay I don't know why it won't let me comment on the line itself but line 108 has a typo in "Parameters"

File: tests/playerclass/test_multiclass.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+#include "common/ctx.h"
+#include "game-state/game.h"
+#include "playerclass/class.h"
+#include "playerclass/class_prefabs.h"
+#include "skilltrees/skilltree.h"
+#include "skilltrees/inventory.h"
+#include "common/utlist.h"
+#include "playerclass/multiclass.h"
+
+//testing here is modeled after test_class_prefabs.c
```
Comment (vanolson) [inline tests/playerclass/test_multiclass.c]: I'm not sure what our policies are on `//` vs `/* */`, but that might be worth changing, and maybe also capitalize testing and put a space before the comment starts

File: tests/playerclass/test_multiclass.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+/* Initializes a chicventure context */
```
Comment (vanolson) [inline tests/playerclass/test_multiclass.c]: typo in chiventure

**Review by vanolson — CHANGES_REQUESTED** [review]
This looks great! In multiclass.c there are a number of lines past 80 characters, which I understand due to all the struct calling, but some of them could be broken up between function arguments and put on new lines to limit the length of the lines. There's also a function definition with a lot of parameters that goes over 80 lines, but I'm not sure if that's something that should be split up. I also noticed a few typos in the file, which I commented. But the tests look very comprehensive, they cover a lot of scenarios. And the functions in multiclass.c look well written. Aside from the few formatting things I commented, this seems to be in great shape!

File: src/playerclass/src/multiclass.c [hunk 4]
```diff
@@ -27,21 +27,77 @@
...
+ 
+    int num_multiclass = base_class->num_parent_class + second_class->num_parent_class;
+
     char* new_shortdesc = (char*) malloc(MAX_SHORT_DESC_LEN + 1);
-    strncat(new_shortdesc, "Multiclass of ", 15);
+    *succ = SUCCESS;
+    int len = 0;
+
+    if ((strstr(base_class->name, "Multiclass of ") == NULL)){
+      len += 14;
```
Comment (lilyehsani) [inline src/playerclass/src/multiclass.c]: Where did these constants come from? These are called [magic numbers](https://uchicago-cs.github.io/dev-guide/style_guide_c.html#avoid-magic-numbers) and they are generally to be avoided. If you can explain in a comment what their purpose is, we can probably allow it. (The numbers I see are 14, 2, 5, and 15).

File: src/playerclass/src/multiclass.c [hunk 5]
```diff
@@ -27,21 +27,77 @@
...
+        return base_class->shortdesc;
+    }
+
+    if ((strstr(base_class->name, "Multiclass of ") == NULL)){
+      strncat(new_shortdesc, "Multiclass of ", 15);
+    }
     strncat(new_shortdesc, base_class->name, strlen(base_class->name));
-    strncat(new_shortdesc, ", ", 3);
+    if (num_multiclass == 2){
+        strncat(new_shortdesc, ", ", 3);
```
Comment (lilyehsani) [inline src/playerclass/src/multiclass.c:82]: (The purpose of 3 and 6 is obvious enough that you do not need to add a comment to these).

File: src/playerclass/src/multiclass.c [hunk 6]
```diff
@@ -27,21 +27,77 @@
  * Returns:
  *  - a pointer to a string with the new shortdesc.
  */
-char* multiclass_shortdesc(class_t* base_class, class_t* second_class) {
+char* multiclass_shortdesc(class_t* base_class, class_t* second_class, int *succ) {
```
Comment (lilyehsani) [inline src/playerclass/src/multiclass.c:31]: Add "succ" as a parameter in the above comment.

File: src/playerclass/src/multiclass.c [hunk 7]
```diff
@@ -56,11 +112,21 @@ char* multiclass_shortdesc(class_t* base_class, class_t* second_class) {
  * Returns:
  *  - a pointer to a string with the new longdesc.
  */
-char* multiclass_longdesc(class_t* base_class, class_t* second_class) {
+char* multiclass_longdesc(class_t* base_class, class_t* second_class, int *succ) {
```
Comment (lilyehsani) [inline src/playerclass/src/multiclass.c:121]: Add "succ" as a parameter in the above comment.

File: src/playerclass/src/multiclass.c (hunk 4 above)
Comment (lilyehsani) [reply src/playerclass/src/multiclass.c]: I see now that they are likely the lengths of strings you are concatenating. Please still add comments to make it clear (for example, `/* 15 = strlen("Multiclass of ") + 1 for the null terminator */`.

File: tests/playerclass/test_class_prefabs.c [hunk 8]
```diff
@@ -45,7 +45,7 @@ chiventure_ctx_t* init_statless_context() {
 }
 
 /* Checks whether the class and its basic fields are not null */
-void check_field_presence(class_t* c) 
+void check_field_pressence(class_t* c)
```
Comment (lilyehsani) [inline tests/playerclass/test_class_prefabs.c:48]: The word "presence" has one s. Please leave this named how it was before to avoid the typo.

**Review by lilyehsani — CHANGES_REQUESTED** [review]
Great job! Just a few comments.

File: tests/playerclass/test_class_prefabs.c (hunk 8 above)
Comment (wellingtj) [reply tests/playerclass/test_class_prefabs.c:48]: it was actually named along the typo before. It was when we fixed it that we got an error

**Review by vanolson — APPROVED** [review]
All my suggestions are fixed, looks great!

**Comment by wellingtj** [comment]
In this PR, we have done some major debugging and performance fixing in the multiclass_shortdesc and multiclass_longdesc functions. We have also implemented complete tests for them

**Review by lilyehsani — APPROVED** [review]
LGTM! Approved.

**Comment by wellingtj** [comment]
> In this PR, we have done some major debugging and performance fixing in the multiclass_shortdesc and multiclass_longdesc functions. We have also implemented complete tests for them

this is the closing statement, jumped the gun on it a bit there

**Comment by lilyehsani** [comment]
Issue Score: Excellent

Comments:
Great work on this PR!

**Pull request merged**